csrc/id_model/predicate_indexing.cpp

```diff
@@ -0,0 +1,56 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <device_lower/utils.h>
+#include <id_model/indexing_utils.h>
+#include <id_model/predicate_indexing.h>
+
+namespace nvfuser {
+
+std::vector<IterDomain*> getPredicateDomains(
+    TensorView* consumer_tv,
+    const Expr* expr) {
+  // Logical domains should be the domains to predicate as they define
+  // the logical shape of a tensor. However, in the case of rfactored
+  // reductions, rfactor splits may not be divisible, thus root
+  // domains need to be predicated. Note that the non-divisible split
+  // info does not seem to cover non-divisible reduction rfactor
+  // splits.
+  std::vector<IterDomain*> predicate_domains = consumer_tv->hasReduction()
+      ? consumer_tv->getMaybeRootDomain()
+      : consumer_tv->getLogicalDomain();
+
+  // Broadcast domains should not need to be predicated. Note that
+  // unlike indexing for TensorIndex, reduction doamins do need to be
+  // indexed to guard the access to the producer tensor
+  predicate_domains.erase(
+      std::remove_if(
+          predicate_domains.begin(),
+          predicate_domains.end(),
+          [](IterDomain* id) -> bool { return id->isBroadcast(); }),
+      predicate_domains.end());
+
+  // If this is an expr initializing a buffer for a reduction, the
+  // reduction domains do not need to be predicated. In fact, if it's
+  // a Local tensor, no predicate is necessary at all
```

Comment:
Why are predicates not needed for Local tensors only? Rather, if there is a reduction domain on a Shared or Global tensor, I don't yet see why they would need predication.

Reply:
If it's a global tensor, since we usually allocate a global memory tensor with its logical domain, the loop domain may be larger than the logical domain if it has non-divisible splits. So, there may be multiple threads initializing the same location if no predication is done.

I think the same issue could happen with shared memory as well. The shape of blockDim and gridDim should be the largest of all the parallelized domains, which may be larger than the parallelized domains of the shared memory tensor. In that case, proper predicates would be needed.

This shouldn't matter in practice, though, since we shouldn't reduce to shared or global memory tensors directly.

Reply:
Ah ok. That makes sense. Thanks for the explanation.